Unity C# version not showing (failed web request)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### What happened?

The C# version isn't showing in the latest release but works fine when I try in the devcontainer. The debug log shows an error when the web request is made to fetch the Unity documentation webpage.

```
context deadline exceeded
```

 Any ideas?

### Theme

Stripped down version of jandedobbeleer.

### What OS are you seeing the problem on?

Windows

### Which shell are you using?

powershell

### Log output

```Shell
[DEBUG] 21:46:56.156 shell.go:Getenv:377 → C:\Users\colt45\AppData\Local
[TRACE] 21:46:56.156 shell.go:Getenv(LOCALAPPDATA) - 0s
[TRACE] 21:46:56.156 shell_windows.go:CachePath() - 0s
[DEBUG] 21:46:56.158 shell.go:Getenv:377 → C:\github\prompt\jandedobbeleer.omp.json
[TRACE] 21:46:56.158 shell.go:Getenv(POSH_THEME) - 0s
[DEBUG] 21:46:56.165 shell.go:Shell:622 → process name: pwsh.exe
[TRACE] 21:46:56.165 shell.go:Shell() - 7.2241ms
[TRACE] 21:46:56.165 shell.go:resolveConfigPath() - 7.2241ms
[TRACE] 21:46:56.165 shell.go:Init() - 8.8573ms
[TRACE] 21:46:56.165 shell.go:Flags() - 0s
[TRACE] 21:46:56.166 config.go:loadConfig() - 650.3µs
[TRACE] 21:46:56.166 shell.go:Flags() - 0s
[DEBUG] 21:46:56.166 shell.go:Getenv:377 → NO DATA
[TRACE] 21:46:56.166 shell.go:Getenv(OMP_CACHE_DISABLED) - 0s
[TRACE] 21:46:56.166 shell_windows.go:WindowsRegistryKeyValue(HKEY_CURRENT_USER\Software\Microsoft\Windows\DWM\ColorizationColor) - 0s
[TRACE] 21:46:56.166 shell.go:GOOS() - 0s
[TRACE] 21:46:56.166 shell.go:GOOS() - 0s
[TRACE] 21:46:56.166 shell.go:GOOS() - 0s
[TRACE] 21:46:56.166 shell.go:GOOS() - 0s
[TRACE] 21:46:56.168 shell.go:GOOS() - 0s
[TRACE] 21:46:56.168 shell.go:GOOS() - 0s
[TRACE] 21:46:56.168 shell.go:GOOS() - 0s
[TRACE] 21:46:56.168 shell.go:GOOS() - 0s
[TRACE] 21:46:56.168 shell.go:GOOS() - 0s
[TRACE] 21:46:56.168 shell.go:GOOS() - 0s
[TRACE] 21:46:56.168 shell.go:GOOS() - 0s
[TRACE] 21:46:56.168 shell.go:GOOS() - 0s
[TRACE] 21:46:56.168 shell.go:GOOS() - 0s
[TRACE] 21:46:56.168 shell.go:GOOS() - 0s
[TRACE] 21:46:56.168 shell.go:GOOS() - 0s
[TRACE] 21:46:56.168 shell.go:GOOS() - 0s
[TRACE] 21:46:56.168 shell.go:GOOS() - 0s
[TRACE] 21:46:56.168 shell.go:GOOS() - 0s
[TRACE] 21:46:56.168 shell.go:GOOS() - 0s
[DEBUG] 21:46:56.169 shell_windows.go:WindowsRegistryKeyValue:206 → ColorizationColor(DWORD): 0x00632812
[TRACE] 21:46:56.169 shell_windows.go:Root() - 0s
[TRACE] 21:46:56.169 shell.go:Shell() - 0s
[TRACE] 21:46:56.169 shell.go:ErrorCode() - 0s
[TRACE] 21:46:56.169 shell_windows.go:IsWsl() - 0s
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[DEBUG] 21:46:56.169 shell.go:Pwd:405 → NO DATA
[TRACE] 21:46:56.169 shell.go:Pwd() - 0s
[DEBUG] 21:46:56.169 shell_windows.go:56 → C:\Users\colt45
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[DEBUG] 21:46:56.169 shell.go:User:526 → colt45
[TRACE] 21:46:56.169 shell.go:User() - 0s
[DEBUG] 21:46:56.169 shell.go:Host:538 → COLT-DEV-DELL
[TRACE] 21:46:56.169 shell.go:Host() - 0s
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[TRACE] 21:46:56.169 shell.go:TemplateCache() - 0s
[TRACE] 21:46:56.169 shell.go:Flags() - 0s
[DEBUG] 21:46:56.169 shell.go:Pwd:385 → C:\Users\colt45\Desktop\Dev\Unity\Scratch\AndroidSandbox
[TRACE] 21:46:56.169 shell.go:Pwd() - 0s
[DEBUG] 21:46:56.169 shell_windows.go:56 → C:\Users\colt45
[TRACE] 21:46:56.169 shell.go:GOOS() - 0s
[DEBUG] 21:46:56.169 shell.go:Getenv:377 → NO DATA
[TRACE] 21:46:56.169 shell.go:Getenv(SSH_CONNECTION) - 0s
[DEBUG] 21:46:56.169 shell.go:Getenv:377 → NO DATA
[TRACE] 21:46:56.169 shell.go:Getenv(SSH_CLIENT) - 0s
[TRACE] 21:46:56.169 shell.go:TemplateCache() - 0s
[TRACE] 21:46:56.170 shell.go:TemplateCache() - 0s
[TRACE] 21:46:56.170 shell.go:Shell() - 0s
[DEBUG] 21:46:56.170 shell.go:Pwd:385 → C:\Users\colt45\Desktop\Dev\Unity\Scratch\AndroidSandbox
[TRACE] 21:46:56.170 shell.go:Pwd() - 0s
[DEBUG] 21:46:56.170 shell_windows.go:56 → C:\Users\colt45
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[DEBUG] 21:46:56.170 shell.go:Pwd:385 → C:\Users\colt45\Desktop\Dev\Unity\Scratch\AndroidSandbox
[TRACE] 21:46:56.170 shell.go:Pwd() - 0s
[TRACE] 21:46:56.170 shell.go:Shell() - 0s
[TRACE] 21:46:56.170 shell.go:Flags() - 0s
[DEBUG] 21:46:56.170 shell_windows.go:56 → C:\Users\colt45
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:PathSeparator() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:PathSeparator() - 0s
[TRACE] 21:46:56.170 shell.go:PathSeparator() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:PathSeparator() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:PathSeparator() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:PathSeparator() - 0s
[TRACE] 21:46:56.170 shell.go:PathSeparator() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:GOOS() - 0s
[TRACE] 21:46:56.170 shell.go:PathSeparator() - 0s
[TRACE] 21:46:56.170 shell.go:PathSeparator() - 0s
[DEBUG] 21:46:56.170 shell.go:Pwd:385 → C:\Users\colt45\Desktop\Dev\Unity\Scratch\AndroidSandbox
[TRACE] 21:46:56.170 shell.go:Pwd() - 0s
[TRACE] 21:46:56.170 shell_windows.go:IsWsl() - 0s
[TRACE] 21:46:56.170 shell.go:StackCount() - 0s
[DEBUG] 21:46:56.171 win32_windows.go:isWriteable:299 → not current user or in group
[DEBUG] 21:46:56.171 win32_windows.go:isWriteable:303 → current user is member of S-1-5-32-544
[DEBUG] 21:46:56.171 win32_windows.go:isWriteable:311 ↓
    WRITE_DAC
    WRITE_OWNER
    SYNCHRONIZE
    DELETE
    READ_CONTROL
[DEBUG] 21:46:56.171 win32_windows.go:isWriteable:313 → user has write access
[TRACE] 21:46:56.171 shell_windows.go:DirIsWritable() - 621.2µs
[TRACE] 21:46:56.171 shell.go:TemplateCache() - 0s
[TRACE] 21:46:56.171 shell.go:TemplateCache() - 0s
[TRACE] 21:46:56.171 shell.go:Shell() - 0s
[DEBUG] 21:46:56.171 shell.go:Pwd:385 → C:\Users\colt45\Desktop\Dev\Unity\Scratch\AndroidSandbox
[TRACE] 21:46:56.171 shell.go:Pwd() - 0s
[DEBUG] 21:46:56.171 shell_windows.go:56 → C:\Users\colt45
[TRACE] 21:46:56.171 shell.go:GOOS() - 0s
[DEBUG] 21:46:56.171 shell.go:Pwd:385 → C:\Users\colt45\Desktop\Dev\Unity\Scratch\AndroidSandbox
[TRACE] 21:46:56.171 shell.go:Pwd() - 0s
[TRACE] 21:46:56.171 shell.go:HasParentFilePath(ProjectSettings) - 0s
[DEBUG] 21:46:56.171 shell.go:HasFilesInDir:436 → true
[TRACE] 21:46:56.171 shell.go:HasFilesInDir(ProjectVersion.txt) - 503.3µs
[DEBUG] 21:46:56.171 shell.go:FileContent:494 ↓
    m_EditorVersion: 2021.3.16f1
    m_EditorVersionWithRevision: 2021.3.16f1 (4016570cf34f)

[TRACE] 21:46:56.171 shell.go:FileContent(C:\Users\colt45\Desktop\Dev\Unity\Scratch\AndroidSandbox\ProjectSettings\ProjectVersion.txt) - 0s
[DEBUG] 21:46:56.171 shell.go:HTTPRequest:666 ↓
    GET /2021.3/Documentation/Manual/CSharpCompiler.html HTTP/1.1
    Host: docs.unity3d.com
    User-Agent: Go-http-client/1.1
    Accept-Encoding: gzip


[ERROR] 21:46:59.178 shell.go:HTTPRequest:670 → Get "https://docs.unity3d.com/2021.3/Documentation/Manual/CSharpCompiler.html": context deadline exceeded
[TRACE] 21:46:59.178 shell.go:HTTPRequest(https://docs.unity3d.com/2021.3/Documentation/Manual/CSharpCompiler.html) - 3.0067786s
[ERROR] 21:46:59.178 unity.go:Enabled:35 → context deadline exceeded
[TRACE] 21:46:59.178 shell.go:TemplateCache() - 0s
[TRACE] 21:46:59.178 shell.go:TemplateCache() - 0s
[TRACE] 21:46:59.179 shell.go:Shell() - 0s
[DEBUG] 21:46:59.179 shell.go:Getenv:377 → C:\Users\colt45\AppData\Local
[TRACE] 21:46:59.179 shell.go:Getenv(LOCALAPPDATA) - 0s
[TRACE] 21:46:59.179 shell_windows.go:CachePath() - 0s
[TRACE] 21:46:59.179 shell.go:Flags() - 0s
```

--- a/src/segments/unity.go
+++ b/src/segments/unity.go
@@ -19,30 +19,6 @@ type Unity struct {
 	CSharpVersion string
 }
 
-var csharpVersionsByUnityVersion = map[string]string{
-	"2017.1": "C# 6",
-	"2017.2": "C# 6",
-	"2017.3": "C# 6",
-	"2017.4": "C# 6",
-	"2018.1": "C# 6",
-	"2018.2": "C# 6",
-	"2018.3": "C# 7.3",
-	"2018.4": "C# 7.3",
-	"2019.1": "C# 7.3",
-	"2019.2": "C# 7.3",
-	"2019.3": "C# 7.3",
-	"2019.4": "C# 7.3",
-	"2020.1": "C# 7.3",
-	"2020.2": "C# 8",
-	"2020.3": "C# 8",
-	"2021.1": "C# 8",
-	"2021.2": "C# 9",
-	"2021.3": "C# 9",
-	"2022.1": "C# 9",
-	"2022.2": "C# 9",
-	"2023.1": "C# 9",
-}
-
 func (u *Unity) Enabled() bool {
 	unityVersion, err := u.GetUnityVersion()
 	if err != nil {
@@ -105,12 +81,36 @@ func (u *Unity) GetCSharpVersion() (version string, err error) {
 	}
 	shortUnityVersion := u.UnityVersion[0:lastDotIndex]
 
+	var csharpVersionsByUnityVersion = map[string]string{
+		"2017.1": "C# 6",
+		"2017.2": "C# 6",
+		"2017.3": "C# 6",
+		"2017.4": "C# 6",
+		"2018.1": "C# 6",
+		"2018.2": "C# 6",
+		"2018.3": "C# 7.3",
+		"2018.4": "C# 7.3",
+		"2019.1": "C# 7.3",
+		"2019.2": "C# 7.3",
+		"2019.3": "C# 7.3",
+		"2019.4": "C# 7.3",
+		"2020.1": "C# 7.3",
+		"2020.2": "C# 8",
+		"2020.3": "C# 8",
+		"2021.1": "C# 8",
+		"2021.2": "C# 9",
+		"2021.3": "C# 9",
+		"2022.1": "C# 9",
+		"2022.2": "C# 9",
+		"2023.1": "C# 9",
+	}
+
 	csharpVersion, found := csharpVersionsByUnityVersion[shortUnityVersion]
 	if found {
 		return csharpVersion, nil
 	}
 
-	u.env.Debug(fmt.Sprintf("Unity version %s doesn't exist in the map. Now making web request...", shortUnityVersion))
+	u.env.Debug(fmt.Sprintf("Unity version %s doesn't exist in the map", shortUnityVersion))
 	return u.GetCSharpVersionFromWeb(shortUnityVersion)
 }
 

--- a/src/segments/unity.go
+++ b/src/segments/unity.go
@@ -19,6 +19,30 @@ type Unity struct {
 	CSharpVersion string
 }
 
+var csharpVersionsByUnityVersion = map[string]string{
+	"2017.1": "C# 6",
+	"2017.2": "C# 6",
+	"2017.3": "C# 6",
+	"2017.4": "C# 6",
+	"2018.1": "C# 6",
+	"2018.2": "C# 6",
+	"2018.3": "C# 7.3",
+	"2018.4": "C# 7.3",
+	"2019.1": "C# 7.3",
+	"2019.2": "C# 7.3",
+	"2019.3": "C# 7.3",
+	"2019.4": "C# 7.3",
+	"2020.1": "C# 7.3",
+	"2020.2": "C# 8",
+	"2020.3": "C# 8",
+	"2021.1": "C# 8",
+	"2021.2": "C# 9",
+	"2021.3": "C# 9",
+	"2022.1": "C# 9",
+	"2022.2": "C# 9",
+	"2023.1": "C# 9",
+}
+
 func (u *Unity) Enabled() bool {
 	unityVersion, err := u.GetUnityVersion()
 	if err != nil {
@@ -76,8 +100,17 @@ func (u *Unity) GetCSharpVersion() (version string, err error) {
 	}
 	shortUnityVersion := u.UnityVersion[0:lastDotIndex]
 
-	if val, found := u.env.Cache().Get(shortUnityVersion); found {
-		csharpVersion := strings.TrimSuffix(val, ".0")
+	csharpVersion, found := csharpVersionsByUnityVersion[shortUnityVersion]
+	if found {
+		return csharpVersion, nil
+	}
+
+	u.env.Debug(fmt.Sprintf("Unity version %s doesn't exist in the map. Now making web request...", shortUnityVersion))
+	return u.GetCSharpVersionFromWeb(shortUnityVersion)
+}
+
+func (u *Unity) GetCSharpVersionFromWeb(shortUnityVersion string) (version string, err error) {
+	if csharpVersion, found := u.env.Cache().Get(shortUnityVersion); found {
 		return csharpVersion, nil
 	}
 

--- a/src/segments/unity.go
+++ b/src/segments/unity.go
@@ -90,7 +90,12 @@ func (u *Unity) GetUnityVersion() (version string, err error) {
 	unityVersion := firstLine[versionStartIndex:]
 	unityVersion = strings.TrimSpace(unityVersion)
 
-	return strings.TrimSuffix(unityVersion, "f1"), nil
+	fIndex := strings.Index(unityVersion, "f")
+	if versionPrefixIndex > -1 {
+		unityVersion = unityVersion[:fIndex]
+	}
+
+	return unityVersion, nil
 }
 
 func (u *Unity) GetCSharpVersion() (version string, err error) {

--- a/src/segments/unity_test.go
+++ b/src/segments/unity_test.go
@@ -36,99 +36,15 @@ func TestUnitySegment(t *testing.T) {
 		Case                string
 		ExpectedOutput      string
 		VersionFileText     string
-		CacheGet            CacheGet
-		CacheSet            CacheSet
 		ExpectedToBeEnabled bool
 		VersionFileExists   bool
-		HTTPResponse        HTTPResponse
 	}{
 		{
-			Case:                "C# version cached",
+			Case:                "Unity version exists in C# map",
 			ExpectedOutput:      "\ue721 2021.3.16 C# 9",
 			ExpectedToBeEnabled: true,
 			VersionFileExists:   true,
 			VersionFileText:     "m_EditorVersion: 2021.3.16f1\nm_EditorVersionWithRevision: 2021.3.16f1 (4016570cf34f)",
-			CacheGet: CacheGet{
-				key:   "2021.3",
-				val:   "C# 9",
-				found: true,
-			},
-		},
-		{
-			Case:                "C# version not cached",
-			ExpectedOutput:      "\ue721 2021.3.16 C# 9",
-			ExpectedToBeEnabled: true,
-			VersionFileExists:   true,
-			VersionFileText:     "m_EditorVersion: 2021.3.16f1\nm_EditorVersionWithRevision: 2021.3.16f1 (4016570cf34f)",
-			CacheGet: CacheGet{
-				key:   "2021.3",
-				val:   "",
-				found: false,
-			},
-			CacheSet: CacheSet{
-				key: "2021.3",
-				val: "C# 9",
-			},
-			HTTPResponse: HTTPResponse{
-				body: `<a href="https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-9">C# 9.0</a>`,
-				err:  nil,
-			},
-		},
-		{
-			Case:                "C# version has a minor version",
-			ExpectedOutput:      "\ue721 2021.3.16 C# 7.3",
-			ExpectedToBeEnabled: true,
-			VersionFileExists:   true,
-			VersionFileText:     "m_EditorVersion: 2021.3.16f1\nm_EditorVersionWithRevision: 2021.3.16f1 (4016570cf34f)",
-			CacheGet: CacheGet{
-				key:   "2021.3",
-				val:   "",
-				found: false,
-			},
-			CacheSet: CacheSet{
-				key: "2021.3",
-				val: "C# 7.3",
-			},
-			HTTPResponse: HTTPResponse{
-				body: `<a href="https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-7-3">C# 7.3</a>`,
-				err:  nil,
-			},
-		},
-		{
-			Case:                "C# version not found in webpage",
-			ExpectedOutput:      "\ue721 2021.3.16",
-			ExpectedToBeEnabled: true,
-			VersionFileExists:   true,
-			VersionFileText:     "m_EditorVersion: 2021.3.16f1\nm_EditorVersionWithRevision: 2021.3.16f1 (4016570cf34f)",
-			CacheGet: CacheGet{
-				key:   "2021.3",
-				val:   "",
-				found: false,
-			},
-			CacheSet: CacheSet{
-				key: "2021.3",
-				val: "",
-			},
-			HTTPResponse: HTTPResponse{
-				body: `<h1>Sorry... that page seems to be missing!</h1>`,
-				err:  nil,
-			},
-		},
-		{
-			Case:                "http request fails",
-			ExpectedOutput:      "\ue721 2021.3.16",
-			ExpectedToBeEnabled: true,
-			VersionFileExists:   true,
-			VersionFileText:     "m_EditorVersion: 2021.3.16f1\nm_EditorVersionWithRevision: 2021.3.16f1 (4016570cf34f)",
-			CacheGet: CacheGet{
-				key:   "2021.3",
-				val:   "",
-				found: false,
-			},
-			HTTPResponse: HTTPResponse{
-				body: "",
-				err:  errors.New("FAIL"),
-			},
 		},
 		{
 			Case:                "ProjectSettings/ProjectVersion.txt doesn't exist",
@@ -153,10 +69,140 @@ func TestUnitySegment(t *testing.T) {
 			ExpectedToBeEnabled: true,
 			VersionFileExists:   true,
 			VersionFileText:     "m_EditorVersion: 2021.3.16f1\r\nm_EditorVersionWithRevision: 2021.3.16f1 (4016570cf34f)\r\n",
+		},
+	}
+
+	for _, tc := range cases {
+		env := new(mock.MockedEnvironment)
+		env.On("Error", mock2.Anything).Return()
+		env.On("Debug", mock2.Anything)
+
+		err := errors.New("no match at root level")
+		var projectDir *platform.FileInfo
+		if tc.VersionFileExists {
+			err = nil
+			projectDir = &platform.FileInfo{
+				ParentFolder: "UnityProjectRoot",
+				Path:         "UnityProjectRoot/ProjectSettings",
+				IsDir:        true,
+			}
+			env.On("HasFilesInDir", projectDir.Path, "ProjectVersion.txt").Return(tc.VersionFileExists)
+			versionFilePath := filepath.Join(projectDir.Path, "ProjectVersion.txt")
+			env.On("FileContent", versionFilePath).Return(tc.VersionFileText)
+		}
+		env.On("HasParentFilePath", "ProjectSettings").Return(projectDir, err)
+
+		props := properties.Map{}
+		unity := &Unity{}
+		unity.Init(props, env)
+		assert.Equal(t, tc.ExpectedToBeEnabled, unity.Enabled())
+		if tc.ExpectedToBeEnabled {
+			assert.Equal(t, tc.ExpectedOutput, renderTemplate(env, unity.Template(), unity), tc.Case)
+		}
+	}
+}
+
+// 2021.9.20f1 is used in the test cases below as a fake Unity version.
+// As such, it doesn't exist in the predfined map in unity.go. This
+// allows us to test the web request portion of the code, which is the
+// fallback for obtaining a C# version.
+func TestUnitySegmentCSharpWebRequest(t *testing.T) {
+	cases := []struct {
+		Case                string
+		ExpectedOutput      string
+		VersionFileText     string
+		CacheGet            CacheGet
+		CacheSet            CacheSet
+		ExpectedToBeEnabled bool
+		VersionFileExists   bool
+		HTTPResponse        HTTPResponse
+	}{
+		{
+			Case:                "C# version cached",
+			ExpectedOutput:      "\ue721 2021.9.20 C# 10",
+			ExpectedToBeEnabled: true,
+			VersionFileExists:   true,
+			VersionFileText:     "m_EditorVersion: 2021.9.20f1\nm_EditorVersionWithRevision: 2021.9.20f1 (4016570cf34f)",
 			CacheGet: CacheGet{
-				key:   "2021.3",
-				val:   "C# 9",
+				key:   "2021.9",
+				val:   "C# 10",
 				found: true,
+			},
+		},
+		{
+			Case:                "C# version not cached",
+			ExpectedOutput:      "\ue721 2021.9.20 C# 10",
+			ExpectedToBeEnabled: true,
+			VersionFileExists:   true,
+			VersionFileText:     "m_EditorVersion: 2021.9.20f1\nm_EditorVersionWithRevision: 2021.9.20f1 (4016570cf34f)",
+			CacheGet: CacheGet{
+				key:   "2021.9",
+				val:   "",
+				found: false,
+			},
+			CacheSet: CacheSet{
+				key: "2021.9",
+				val: "C# 10",
+			},
+			HTTPResponse: HTTPResponse{
+				body: `<a href="https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-10">C# 10.0</a>`,
+				err:  nil,
+			},
+		},
+		{
+			Case:                "C# version has a minor version",
+			ExpectedOutput:      "\ue721 2021.9.20 C# 10.1",
+			ExpectedToBeEnabled: true,
+			VersionFileExists:   true,
+			VersionFileText:     "m_EditorVersion: 2021.9.20f1\nm_EditorVersionWithRevision: 2021.9.20f1 (4016570cf34f)",
+			CacheGet: CacheGet{
+				key:   "2021.9",
+				val:   "",
+				found: false,
+			},
+			CacheSet: CacheSet{
+				key: "2021.9",
+				val: "C# 10.1",
+			},
+			HTTPResponse: HTTPResponse{
+				body: `<a href="https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-10-1">C# 10.1</a>`,
+				err:  nil,
+			},
+		},
+		{
+			Case:                "C# version not found in webpage",
+			ExpectedOutput:      "\ue721 2021.9.20",
+			ExpectedToBeEnabled: true,
+			VersionFileExists:   true,
+			VersionFileText:     "m_EditorVersion: 2021.9.20f1\nm_EditorVersionWithRevision: 2021.9.20f1 (4016570cf34f)",
+			CacheGet: CacheGet{
+				key:   "2021.9",
+				val:   "",
+				found: false,
+			},
+			CacheSet: CacheSet{
+				key: "2021.9",
+				val: "",
+			},
+			HTTPResponse: HTTPResponse{
+				body: `<h1>Sorry... that page seems to be missing!</h1>`,
+				err:  nil,
+			},
+		},
+		{
+			Case:                "http request fails",
+			ExpectedOutput:      "\ue721 2021.9.20",
+			ExpectedToBeEnabled: true,
+			VersionFileExists:   true,
+			VersionFileText:     "m_EditorVersion: 2021.9.20f1\nm_EditorVersionWithRevision: 2021.9.20f1 (4016570cf34f)",
+			CacheGet: CacheGet{
+				key:   "2021.9",
+				val:   "",
+				found: false,
+			},
+			HTTPResponse: HTTPResponse{
+				body: "",
+				err:  errors.New("FAIL"),
 			},
 		},
 	}

--- a/website/docs/segments/unity.mdx
+++ b/website/docs/segments/unity.mdx
@@ -8,13 +8,16 @@ sidebar_label: Unity
 
 Display the currently active [Unity][unity] and C# versions.
 
-The Unity version is displayed regardless of whether or not the C# version fetching succeeds.
+The Unity version is displayed regardless of whether or not the corrsponding C# version can be found.
+The C# version is determined by first checking a static table.
+If the Unity version isn't found, a web request is made to [the Unity docs][unity-csharp-page] to
+try extracting it from there. A web request only occurs the first time a given `major.minor`
+Unity version is encountered. Subsequent invocations return the cached C# version.
 
-The C# version is fetched from [the docs][unity-csharp-page], but the page only exists as far back as Unity 2018.3.
-Earlier versions will not display a C# version.
+C# version display is only supported from Unity 2017.1.
 
-A web request only occurs the first time a given `major.minor` Unity version is encountered.
-Subsequent invocations return the cached C# version.
+Unity 2017.1 - 2019.1 support two C# versions, depending on which scripting runtime is selected in Player Settings.
+This segment always chooses the higher version.
 
 ## Sample Configuration
 

